### DISCUSSION TITLE
Why are scans so much slower in v2 compared to v1?

### DISCUSSION THREAD

**Discussion by gparonitti** [dynamoose/dynamoose#1085] [Q&A]
Hi there, we have a project with several deployment environments. Recently we have updated in the development env the Dynamoose version to 2.x (updating the models accordingly). All the nodejs packages reside on a lambda layer. We have noticed that the scan is 5x times slower than the v1.x which lingers in production. We are going crazy, in figuring out why this decrease in performance (on the same data set of ~500 items we pass from 1.5-2 sec in production with version 1.x to 8-11 secs in development with version 2.x)

Dev scan all result:
![image](https://user-images.githubusercontent.com/11628507/102793075-a9121280-43a9-11eb-84ca-f209f287f524.png)

Prod scan all result:
![image](https://user-images.githubusercontent.com/11628507/102794335-88e35300-43ab-11eb-8140-4e043fb3cc01.png)

Has anyone had or noticed the same problem?

**Comment by fishcharlie** [comment]
I'm 100% aware that some operations in v2 are slower. This is in large part to the more ￼thorough type checking that occurs. However there could be other factors at play here.

The changes made in v2 I believe were the correct changes. It fixed a lot of bugs and made the entire system more reliable. So I standby it even tho it has these side effects.

Sadly, I'm not an expert at solving performance issues. I know the basics, and have tried to optimize where I can, but it's always a trade off between speed and resources (memory usage, etc).

Ideally we'd work towards improving this by doing the following:

1. Write a system for doing performance checks. I don't wanna make a bunch of changes just to have future work mess it up.
2. Figure out where specifically is the worst performance code blocks and address those first. Not sure how we can test where this problem is occurring.

The big question for item 2 is, how do we type check 700+ items in the fastest way possible? Of course it scales with each item and the more attributes you have as well. 700 items with 1 attribute is gonna go much faster than 700 attributes on 700 items. Not sure if it'd go faster or slower than 1 item with 700 attributes tho.

Definitely questions that will need to be investigated.

I'd love to chat about this more. This is the type of project that is slightly over my head but I do recognize it's importance. We are lucky enough to have Dynobase as a sponsor, so it's likely if someone wants to work on solving this, I'd be able to give them a small cash bonus or something. Nothing major, but it'd at least provide a little incentive.

---

Totally unrelated (and likely wouldn't be eligible for incentives) would be adding a command to bypass type checking.

That wouldn't actually solve the problem and would be a band-aid fix. But it might be an interesting concept to think about and consider more. Would love to hear input on this.

---

I'm sorry I can't provide a more concrete answer. And I really wish this wasn't the case. Hopefully we can get some interested towards making this a priority.

**Reply by gparonitti** [reply]
Thank you for the answer and for the great job you provide with Dynamoose. We moved our infrastructure to AWS one year ago, with a good experience in MongoDB and Mongoose, so having found Dynamoose has given us a very good help in the process (even if we realized that we were overlooking some internal mechanisms of DynamoDB, which seems more a multidimensional Redis than a typical noSQL).

Regarding the issue: we already do some type checking with fastest-validator before saving, so we are pretty sure that data are correct at the source and, let's say, inside the DDB table. Would not be feasible to let the type checking optional for scans? And is the type checking really needed on getting the items from the table?

Generally speaking, one of the main reasons of using DynamoDB is that it is insanely fast (despite its complexity in indexing data to query them so fast). Thus, it seems a priority to have a middleware that respects this kind of performances. In my understanding the model validation is a burden of the developer especially for a db engine where the data types are arguably irrelevant, so that the developer should do all the legs work to cope with this matter.

**Reply by fishcharlie** [reply]
@gparonitti

> for the great job you provide with Dynamoose
> having found Dynamoose has given us a very good help in the process

So glad to hear this!! 😃 

> we already do some type checking with fastest-validator before saving

Not quite sure what you mean by this. If you are doing type checking on your own (without Dynamoose), then that applies likely just to your use case. Probably doesn't apply to the majority of Dynamoose users.

> so we are pretty sure that data are correct at the source and, let's say, inside the DDB table

I've for sure seen cases from other users that can't say the same. Very easy for things to go wrong, and Dynamoose alerting you can be a huge benefit.

> Would not be feasible to let the type checking optional for scans? And is the type checking really needed on getting the items from the table?

So this is kinda what I mentioned at the end of my response. It sounds easy, but there is a lot of complexity here. Just some first things that come to mind:

1. Need to ensure this works for all read operations (get, batchGet, query, etc).
2. Need to come up with a good name for this setting
3. Currently the type checking method is tied in with other actions such as the [`get` setting](https://dynamoosejs.com/guide/Schema#get-function--async-function). It also is tied into the [Custom Dynamoose Types](https://dynamoosejs.com/guide/Schema#attribute-types). So we'd need to figure out how to allow those things.

One thing I forget, is if Mongoose does type checking when retrieving items. Doing an investigation into that would be good to provide us with more information.

I also still believe this would be a band-aid fix. And the likelihood that it reduces it to v1 levels is 50/50 I'd say.

> one of the main reasons of using DynamoDB is that it is insanely fast (despite its complexity in indexing data to query them so fast). Thus, it seems a priority to have a middleware that respects this kind of performances

Totally agree. As I said: `I do recognize it's importance`.

> In my understanding the model validation is a burden of the developer especially for a db engine where the data types are arguably irrelevant, so that the developer should do all the legs work to cope with this matter

I disagree with this. One of the reasons users use Dynamoose is to provide that for them. I think most developers who use Dynamoose would not like the burden of doing type checking.

---

One other thing I forgot to mention. As part of v2, we are now using the AWS-SDK to parse DynamoDB items into standard JSON objects. Through a little bit of investigation previously, that process takes an EXTREMELY long time. It might be worth trying to build a system to speed that up, and use our own code to handle that parsing again. It'd require a lot of testing to determine which system is faster tho, and if there are any side effects with that.

However, I'm aware that is not the only thing impacting speed right now. I think in some cases we are getting the details of the schema attribute multiple times when doing those checks which slows things down. Figuring out a way to cache that somehow maybe in an object might speed things up.

---

If you could. Would you mind doing one more test for me?

1. In v2, change your schema to **only** have your hashKey and rangeKey (if exists). So like reduce your schema to be as small as possible. Ideally that means removing a **lot** of attributes that are adding to the complexity of your type checking. I'm not sure what your schema looks like, so if you are only able to remove one or two, this test won't be as useful. If you are able to remove a lot however, this test will be useful.
2. Set `saveUnknown` to `true` in your Schema settings.
3. Run the same test again, and screenshot your results and post here.

That should speed things up quite a bit. It should be bypassing a lot of the complexity that occurs. There is still some things it does, but it should streamline things.

Curious on what order of magnitude that changes things tho.
